// @web-and-design/eslint-plugin-wed/lib/rules/use-action-map.js
"use strict";

//------------------------------------------------------------------------------
// Rule Definition
//------------------------------------------------------------------------------

module.exports = {
    meta: {
        docs: {
            description: "use action map instead of else if or switch case",
            category: "testat",
            recommended: true
        },
        fixable: null,
        schema: []
    },
    create: function (context) {

        return {
            IfStatement(node) {
                if (node) {
                    if (node.alternate && node.alternate.type === "IfStatement" && node.parent.type !== "IfStatement") {
                        context.report({
                            node,
                            message: "Use action map instead if - else if - else if.",
                            type: "problem"
                        });
                    }
                    // node.parent.body must be null otherwise gets trigged by else if
                    else if (node.parent != null && node.parent.body != null) {
                        const index = node.parent.body.indexOf(node);
                        const nextNode = node.parent.body[index + 1];
                        const prevNode = node.parent.body[index - 1];
                        if (nextNode != null && nextNode.type === "IfStatement" && !(prevNode != null && prevNode.type === "IfStatement")) {
                            context.report({
                                node,
                                message: "Use action map instead if - if - ...",
                                type: "problem"
                            });
                        }
                    }
                }
            },
            SwitchStatement(node) {
                if (node) {

                    context.report({
                        node,
                        message: "Use action map instead if switch case.",
                        type: "problem"
                    });
                }
            }
        };
    }
};
